src/main/java/net/mcreator/element/types/Recipe.java

```diff
@@ -137,44 +137,48 @@ public Recipe(ModElement element) {
 		return mod;
 	}
 
-	private final transient OptimisedRecipe optimisedRecipe;
-
 	public MItemBlock[][] getOptimisedRecipe() {
-		return optimisedRecipe.getOptimisedRecipe();
-	}
-
-	private record OptimisedRecipe(Recipe recipe) {
-
-		MItemBlock[][] getOptimisedRecipe() {
-			MItemBlock[][] mtx = { { recipe.recipeSlots[0], recipe.recipeSlots[1], recipe.recipeSlots[2] },
-					{ recipe.recipeSlots[3], recipe.recipeSlots[4], recipe.recipeSlots[5] },
-					{ recipe.recipeSlots[6], recipe.recipeSlots[7], recipe.recipeSlots[8] } };
-			int cmin = mtx[0].length;
-			int rmin = mtx.length;
-			int cmax = -1;
-			int rmax = -1;
-			for (int r = 0; r < mtx.length; r++)
-				for (int c = 0; c < mtx[0].length; c++)
-					if (!mtx[r][c].isEmpty()) {
-						if (cmin > c)
-							cmin = c;
-						if (cmax < c)
-							cmax = c;
-						if (rmin > r)
-							rmin = r;
-						if (rmax < r)
-							rmax = r;
-					}
-			return trim(mtx, rmin, rmax, cmin, cmax);
+		MItemBlock[][] mtx = { { recipeSlots[0], recipeSlots[1], recipeSlots[2] },
+				{ recipeSlots[3], recipeSlots[4], recipeSlots[5] },
+				{ recipeSlots[6], recipeSlots[7], recipeSlots[8] } };
+
+		// determine recipe matrix edges
+		int cmin = mtx[0].length;
+		int rmin = mtx.length;
+		int cmax = -1;
+		int rmax = -1;
+		for (int r = 0; r < mtx.length; r++) {
+			for (int c = 0; c < mtx[0].length; c++) {
+				if (!mtx[r][c].isEmpty()) {
+					if (cmin > c)
+						cmin = c;
+					if (cmax < c)
+						cmax = c;
+					if (rmin > r)
+						rmin = r;
+					if (rmax < r)
+						rmax = r;
+				}
+			}
 		}
 
-		private MItemBlock[][] trim(MItemBlock[][] mtx, int rmin, int rmax, int cmin, int cmax) {
-			MItemBlock[][] result = new MItemBlock[rmax - rmin + 1][];
-			for (int r = rmin, i = 0; r <= rmax; r++, i++)
-				result[i] = Arrays.copyOfRange(mtx[r], cmin, cmax + 1);
-			return result;
-		}
+		// trim array size to the smallest possible
+		MItemBlock[][] result = new MItemBlock[rmax - rmin + 1][];
+		for (int r = rmin, i = 0; r <= rmax; r++, i++)
+			result[i] = Arrays.copyOfRange(mtx[r], cmin, cmax + 1);
+		return result;
+	}
 
+	public Map<MItemBlock, String> getPatternKeys() {

```

Comment:
Why would you build the map the other way around? (Feels more natural if keys were strings)